File: scripts/deg_cell_type_vs_rest_nonpb.py
"""
deg_cell_type_vs_rest_nonpb.py

Differential expression: one cell type vs. all other cells (not pseudobulked).

Uses scanpy's `rank_genes_groups` with the Wilcoxon rank-sum test, which is the
closest equivalent to Seurat's FindAllMarkers (wilcox method).

We run the test directly on the single-cell level (no pseudobulking), keeping the
expression matrix sparse throughout to avoid OOM errors.

Output CSV columns
------------------
gene            gene name / ID
score           Wilcoxon z-score
log2fc          log2 fold change (log1p_norm scale)
pval            uncorrected p-value
pval_adj        Benjamini-Hochberg adjusted p-value
n_cells_fg      number of cells in the foreground (cell type of interest)
n_cells_bg      number of cells in the background (all other cells)
n_donors_fg     number of unique donors in the foreground
n_donors_bg     number of unique donors in the background
"""

import argparse
import os
import warnings

import numpy as np
import pandas as pd
import scanpy as sc
from anndata import read_h5ad

DONOR_COL = "patient"


def unsanitize(sanitized: str, candidates) -> str:
    for c in candidates:
        if c.replace(" ", "_").replace("/", "_") == sanitized:
            return c
    return sanitized


def run_deg(adata: sc.AnnData, cell_type_col: str, cell_type: str) -> pd.DataFrame:
    # Binary grouping: foreground = cell type of interest, background = rest
    fg_mask = adata.obs[cell_type_col] == cell_type
    bg_mask = ~fg_mask

    n_fg = fg_mask.sum()
    n_bg = bg_mask.sum()
    n_donors_fg = adata.obs.loc[fg_mask, DONOR_COL].nunique() if DONOR_COL in adata.obs.columns else np.nan
    n_donors_bg = adata.obs.loc[bg_mask, DONOR_COL].nunique() if DONOR_COL in adata.obs.columns else np.nan

    print(f"  Foreground: {n_fg} cells ({n_donors_fg} donors), Background: {n_bg} cells ({n_donors_bg} donors)")

    if n_fg == 0:
        raise ValueError(f"No cells found for cell type '{cell_type}' in column '{cell_type_col}'.")

    # Assign binary labels for rank_genes_groups
    adata.obs["_group"] = "rest"
    adata.obs.loc[fg_mask, "_group"] = "target"

    # Use log1p_norm layer if available (normalized + log-transformed), else X
    if "log1p_norm" in adata.layers:
        adata_tmp = adata.copy()
        adata_tmp.X = adata_tmp.layers["log1p_norm"]
    else:
        adata_tmp = adata

    with warnings.catch_warnings():
        warnings.simplefilter("ignore")
        sc.tl.rank_genes_groups(
            adata_tmp,
            groupby="_group",
            groups=["target"],
            reference="rest",
            method="wilcoxon",
            use_raw=False,
            pts=True,
        )

    result = sc.get.rank_genes_groups_df(adata_tmp, group="target")
    result = result.rename(
        columns={
            "names": "gene",
            "scores": "score",
            "logfoldchanges": "log2fc",
            "pvals": "pval",
            "pvals_adj": "pval_adj",
        }
    )

    result["n_cells_fg"] = int(n_fg)
    result["n_cells_bg"] = int(n_bg)
    result["n_donors_fg"] = int(n_donors_fg) if not np.isnan(n_donors_fg) else np.nan
    result["n_donors_bg"] = int(n_donors_bg) if not np.isnan(n_donors_bg) else np.nan
    result["cell_type"] = cell_type
    result["cell_type_col"] = cell_type_col

    # Drop the temporary grouping column from adata.obs
    adata.obs.drop(columns=["_group"], inplace=True, errors="ignore")

    return result


def main():
    parser = argparse.ArgumentParser(
        description="Cell-type-vs-rest DEG (not pseudobulked, Wilcoxon)."
    )
    parser.add_argument("--input", required=True, help="Normalized H5AD path.")
    parser.add_argument("--output", required=True, help="Output CSV path.")
    parser.add_argument("--cell-type-col", required=True)
    parser.add_argument("--cell-type", required=True)
    args = parser.parse_args()

    os.makedirs(os.path.dirname(args.output), exist_ok=True)

    print(f"Loading: {args.input}")
    adata = read_h5ad(args.input)

    actual_cell_type = unsanitize(
        args.cell_type, adata.obs[args.cell_type_col].unique().tolist()
    )
    if actual_cell_type != args.cell_type:
        print(f"  Resolved cell type '{args.cell_type}' -> '{actual_cell_type}'")

    print(f"Running Wilcoxon DEG for '{actual_cell_type}' vs rest ...")
    df = run_deg(adata, args.cell_type_col, actual_cell_type)

    df.to_csv(args.output, index=False)
    print(f"Wrote {len(df)} rows to {args.output}")


if __name__ == "__main__":
    main()

File: scripts/deg_cell_type_vs_rest_nonpb.R
# deg_cell_type_vs_rest_nonpb.R
#
# Differential expression: one cell type vs. all other cells (not pseudobulked).
#
# Uses Seurat's FindMarkers with the Wilcoxon rank-sum test.
# Input is the normalized H5AD file (produced by normalize.py).
# The log1p_norm layer is used as the expression matrix for the test.
#
# Output CSV columns
# ------------------
# gene            gene name / ID
# avg_log2FC      average log2 fold change (Seurat convention)
# pval            uncorrected p-value  (p_val in Seurat output)
# pval_adj        Bonferroni-corrected p-value  (p_val_adj)
# pct_fg          fraction of cells in foreground expressing the gene  (pct.1)
# pct_bg          fraction of cells in background expressing the gene  (pct.2)
# n_cells_fg      number of cells in the foreground (cell type of interest)
# n_cells_bg      number of cells in the background (all other cells)
# n_donors_fg     number of unique donors in the foreground
# n_donors_bg     number of unique donors in the background
# cell_type       the target cell type
# cell_type_col   the obs column used for grouping

suppressPackageStartupMessages({
  library(anndataR)
  library(Seurat)
  library(dplyr)
})

# ---------------------------------------------------------------------------
# Argument parsing
# ---------------------------------------------------------------------------
args <- commandArgs(trailingOnly = TRUE)

parse_args <- function(args) {
  out <- list()
  i <- 1
  while (i <= length(args)) {
    key <- sub("^--", "", args[i])
    key <- gsub("-", "_", key)  # --cell-type-col -> cell_type_col
    val <- args[i + 1]
    out[[key]] <- val
    i <- i + 2
  }
  out
}

opts <- parse_args(args)

required <- c("input", "output", "cell_type_col", "cell_type")
missing_args <- setdiff(required, names(opts))
if (length(missing_args) > 0) {
  stop(paste("Missing required arguments:", paste(paste0("--", gsub("_", "-", missing_args)), collapse = ", ")))
}

input_path    <- opts$input
output_path   <- opts$output
cell_type_col <- opts$cell_type_col
cell_type_san <- opts$cell_type   # may be sanitized (spaces/slashes replaced)

# ---------------------------------------------------------------------------
# Helper: recover original value from a sanitized wildcard
# ---------------------------------------------------------------------------
unsanitize <- function(sanitized, candidates) {
  normed <- gsub("/", "_", gsub(" ", "_", candidates))
  idx <- which(normed == sanitized)
  if (length(idx) > 0) candidates[idx[1]] else sanitized
}

# ---------------------------------------------------------------------------
# Load H5AD as AnnData first to resolve the cell type label
# ---------------------------------------------------------------------------
cat(sprintf("Loading: %s\n", input_path))
adata <- read_h5ad(input_path)

# Recover actual cell type label (handles VSM/P -> VSM_P etc.)
all_cell_types <- unique(adata$obs[[cell_type_col]])
cell_type <- unsanitize(cell_type_san, all_cell_types)
if (cell_type != cell_type_san) {
  cat(sprintf("  Resolved cell type '%s' -> '%s'\n", cell_type_san, cell_type))
}

donor_col <- "patient"

# Compute counts before converting to Seurat (obs is an R data.frame in anndataR)
obs <- adata$obs
fg_cells <- obs[[cell_type_col]] == cell_type
n_fg <- sum(fg_cells)
n_bg <- sum(!fg_cells)

if (n_fg == 0) {
  stop(sprintf("No cells found for cell type '%s' in column '%s'.", cell_type, cell_type_col))
}

n_donors_fg <- if (donor_col %in% colnames(obs)) {
  length(unique(obs[[donor_col]][fg_cells]))
} else NA_integer_

n_donors_bg <- if (donor_col %in% colnames(obs)) {
  length(unique(obs[[donor_col]][!fg_cells]))
} else NA_integer_

cat(sprintf("  Foreground: %d cells (%s donors), Background: %d cells (%s donors)\n",
            n_fg,
            ifelse(is.na(n_donors_fg), "NA", as.character(n_donors_fg)),
            n_bg,
            ifelse(is.na(n_donors_bg), "NA", as.character(n_donors_bg))))

# ---------------------------------------------------------------------------
# Convert to Seurat, mapping log1p_norm layer as X (-> data slot)
# ---------------------------------------------------------------------------
cat("Converting to Seurat object...\n")

# x_mapping = "log1p_norm" tells as_Seurat to use the log1p_norm layer as X,
# which populates the 'data' slot that FindMarkers (wilcox) reads from.
# We also bring counts along so Seurat is fully formed.
seurat_obj <- adata$as_Seurat(
  x_mapping      = "log1p_norm",
  layers_mapping = c("counts")
)

# ---------------------------------------------------------------------------
# Set identity classes and run FindMarkers
# ---------------------------------------------------------------------------
Idents(seurat_obj) <- seurat_obj@meta.data[[cell_type_col]]

cat(sprintf("Running FindMarkers (Wilcoxon) for '%s' vs rest...\n", cell_type))

markers <- FindMarkers(
  seurat_obj,
  ident.1         = cell_type,
  ident.2         = NULL,   # NULL = all other cells
  test.use        = "wilcox",
  logfc.threshold = 0,      # return all genes, filter later
  min.pct         = 0,
  only.pos        = FALSE,
  verbose         = FALSE
)

# ---------------------------------------------------------------------------
# Tidy output
# ---------------------------------------------------------------------------
result <- markers %>%
  tibble::rownames_to_column("gene") %>%
  rename(
    pval     = p_val,
    pval_adj = p_val_adj,
    pct_fg   = pct.1,
    pct_bg   = pct.2,
  ) %>%
  mutate(
    n_cells_fg    = as.integer(n_fg),
    n_cells_bg    = as.integer(n_bg),
    n_donors_fg   = as.integer(n_donors_fg),
    n_donors_bg   = as.integer(n_donors_bg),
    cell_type     = cell_type,
    cell_type_col = cell_type_col,
  ) %>%
  select(gene, avg_log2FC, pval, pval_adj, pct_fg, pct_bg,
         n_cells_fg, n_cells_bg, n_donors_fg, n_donors_bg,
         cell_type, cell_type_col)

# ---------------------------------------------------------------------------
# Write output
# ---------------------------------------------------------------------------
dir.create(dirname(output_path), recursive = TRUE, showWarnings = FALSE)
write.csv(result, output_path, row.names = FALSE)
cat(sprintf("Wrote %d rows to %s\n", nrow(result), output_path))
